usecase/src/main/java/org/mobilitydata/gtfsvalidator/usecase/ValidateShapeDistTraveled.java

```diff
@@ -71,23 +67,22 @@ public void execute() {
         stopTimePerTripId.forEach(
                 (tripId, stopTimeCollection) -> dataRepo.getStopTimeByTripId(tripId)
                         .forEach((stopSequence, stopTime) -> {
-                            if (stopTime.getShapeDistTraveled() != null) {
+                            final Float shapeDistanceTraveled = stopTime.getShapeDistTraveled();
+                            if (shapeDistanceTraveled != null) {
                                 final Trip trip = dataRepo.getTripById(tripId);
                                 if (trip != null) {
                                     final String shapeId = trip.getShapeId();
                                     if (shapeId != null) {
                                         final Map<Integer, ShapePoint> shape = dataRepo.getShapeById(shapeId);
                                         if (shape != null) {
-                                            if (stopTime.getShapeDistTraveled() >
-                                                    distanceCalculationUtils
-                                                            .getShapeTotalDistance(shape, DistanceUnit.KILOMETER)) {
+                                            final float shapeTotalDistance = (float) distanceCalculationUtils
```

Comment:
A few comments here. If I'm reading this right it seems really inefficient to me. Here's how I read it for each trip:

For each record in stop_times.txt for a given trip_id:
* Calculate the entire shape length using lat/lons from shapes.txt
* Compare that entire shape length against this stop_times.txt shapeDistanceTraveled value

You shouldn't need to recalculate the entire shape distance for each record - that can be computed once per trip before looping through the stop_times.txt records for that trip.

But - reading the spec for shape_dist_traveled more carefully, I don't think this rule is correct. If anything, the shape_dist_traveled may be *more* than the "connect the dots" length for the lat/lons in shapes.txt. My understanding is that typically shape_dist_traveled is calculated using odometers, which means the distance may have a "curve" to it in between points. I haven't really used this field much myself, so I'd like to get feedback from @timMillet if this is correct.

Reply:
Good catch, will modify the implementation. Waiting for Tim's feedback though.

Reply:
The other complication is that the units for shape_dist_traveled aren't specified in the spec, which also makes this impossible to implement.

Reply:
Ok, after going back and reviewing https://github.com/MobilityData/gtfs-validator/issues/163, I don't think this implementation is the correct interpretation of that rule (I looked quickly for the Python implementation but didn't see it, but I asked @maximearmstrong to add a pointer to it in https://github.com/MobilityData/gtfs-validator/issues/163).

Here's my interpretation:

```
For each of the records in shapes.txt for a given trip_id, check that each shape_dist_traveled is less than the maximum shape_dist_traveled for that trip_id.
```

Since our shape data is already sorted, you could further improve on this as:

```
For each of the records in shapes.txt for a given trip_id sorted by shape_pt_sequence, check that each shape_dist_traveled is ascending as shape_pt_sequence increases.
```

Let's review the Python implementation before any changes to make sure we're understanding this correctly.

Reply:
Also, the above logic for the rule can also apply to stop_times.txt data, sorted by stop_sequence.